src/app/services/catalogue/catalogue.service.ts

```diff
@@ -37,16 +37,44 @@ export class CatalogueService {
     const topicFilter = topic ? `+topic:${topic}` : '';
     const orgFilter = org ? `org:${org}+` : '';
 
-    return this.getFromAPI(
-      `https://api.github.com/search/repositories?page=${this.CONF.page}&per_page=${this.CONF.perPage}&q=${orgFilter}fork:true${topicFilter}+sort:stars`,
-      key,
+    const perPage = this.CONF.perPage;
+    let totalSize = null;
+
+    const getPage = (page) => {
+      return this.http.get<RepositoryPage>(`https://api.github.com/search/repositories?page=${page}&per_page=${this.CONF.perPage}&q=${orgFilter}fork:true${topicFilter}+sort:stars`);
+    };
+
+    const pages$ = new Observable<Repository[]>((observer) => {
+      const emitItems = (page) => {
+        getPage(page).pipe(
+          map((res) => {
+            if (!totalSize) {
+              totalSize = res.total_count;
+            }
+            return res.items;
+          }),
+        ).subscribe((items) => {
+          observer.next(items);
+          if (perPage * page < totalSize) {
+            emitItems(page + 1);
+          } else {
+            observer.complete();
+          }
+        });
+      };
+
+      emitItems(1);
+    }).pipe(
+      reduce((acc, val) => acc.concat(val), []),
     );
+
+    return this.storable(pages$, key);
   }
 
-  getFromAPI(URL: string, key: string): Observable<Repository> {
-    return this.http.get(URL).pipe(
+  storable(obs: Observable<any>, key: string): Observable<any> {
```

Comment:
why `any`? is it `Repository`?

Reply:
This function is used by two methods, one which passes a `Repository` and another one passing `Repository[]`. I improved it a little bit using generics.